// zcatalyst-sdk-node/lib/zia/zia.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Zia = void 0;
const api_request_1 = require("../utils/api-request");
const constants_1 = __importDefault(require("../utils/constants"));
const error_1 = require("../utils/error");
const form_data_1 = __importDefault(require("../utils/form-data"));
const validator_1 = require("../utils/validator");
const { REQ_METHOD, COMPONENT } = constants_1.default;
function addOptsToFormData(formData, opts) {
    for (const [key, value] of Object.entries(opts)) {
        if (value !== undefined) {
            formData.append(key, value);
        }
    }
}
class Zia {
    constructor(app) {
        validator_1.isValidApp(app, true);
        this.requester = new api_request_1.AuthorizedHttpClient(app, this);
    }
    getComponentName() {
        return COMPONENT.zia;
    }
    detectObject(file) {
        return __awaiter(this, void 0, void 0, function* () {
            const formData = new form_data_1.default();
            formData.append('image', file);
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/detect-object`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    extractOpticalCharacters(file, opts = {
        language: undefined,
        modelType: undefined
    }) {
        return __awaiter(this, void 0, void 0, function* () {
            const formData = new form_data_1.default();
            formData.append('image', file);
            addOptsToFormData(formData, opts);
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/ocr`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    extractAadhaarCharacters(frontImg, backImg, language) {
        return __awaiter(this, void 0, void 0, function* () {
            yield validator_1.wrapValidatorsWithPromise(() => {
                validator_1.isNonEmptyString(language, 'language', true);
            }, error_1.CatalystZiaError);
            const formData = new form_data_1.default();
            formData.append('aadhaar_front', frontImg);
            formData.append('aadhaar_back', backImg);
            addOptsToFormData(formData, {
                language,
                model_type: 'AADHAAR'
            });
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/ocr`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    /**
     * @deprecated
     */
    extractNamedEntity(file) {
        return __awaiter(this, void 0, void 0, function* () {
            const formData = new form_data_1.default();
            if (validator_1.isNonEmptyString(file)) {
                formData.append('text', file);
            }
            else {
                formData.append('text_file', file);
            }
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/ner`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    scanBarcode(image, opts = { format: undefined }) {
        return __awaiter(this, void 0, void 0, function* () {
            yield validator_1.wrapValidatorsWithPromise(() => {
                validator_1.isValidType(image, 'object', 'image', true);
            }, error_1.CatalystZiaError);
            const formData = new form_data_1.default();
            formData.append('image', image);
            addOptsToFormData(formData, opts);
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/barcode`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    moderateImage(image, opts = { mode: undefined }) {
        return __awaiter(this, void 0, void 0, function* () {
            yield validator_1.wrapValidatorsWithPromise(() => {
                validator_1.isValidType(image, 'object', 'image', true);
            }, error_1.CatalystZiaError);
            const formData = new form_data_1.default();
            formData.append('image', image);
            addOptsToFormData(formData, opts);
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/imagemoderation`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    analyseFace(image, opts = {
        mode: undefined,
        emotion: undefined,
        age: undefined,
        gender: undefined
    }) {
        return __awaiter(this, void 0, void 0, function* () {
            yield validator_1.wrapValidatorsWithPromise(() => {
                validator_1.isValidType(image, 'object', 'image', true);
            }, error_1.CatalystZiaError);
            const formData = new form_data_1.default();
            formData.append('image', image);
            addOptsToFormData(formData, opts);
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/faceanalytics`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    /**
     * @param sourceImage read stream of the source image
     * @param queryImage read stream of the image to be compared
     * @returns {ICatalystZiaFaceComparison} the object that contains the match and confidence value
     */
    compareFace(sourceImage, queryImage) {
        return __awaiter(this, void 0, void 0, function* () {
            yield validator_1.wrapValidatorsWithPromise(() => {
                validator_1.isValidType(sourceImage, 'object', 'source_image', true);
                validator_1.isValidType(queryImage, 'object', 'query_image', true);
            }, error_1.CatalystZiaError);
            const formData = new form_data_1.default();
            formData.append('source_image', sourceImage);
            formData.append('query_image', queryImage);
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/facecomparison`,
                data: formData,
                type: 'file',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    automl(modelId, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            yield validator_1.wrapValidatorsWithPromise(() => {
                validator_1.isNonEmptyStringOrNumber(modelId, 'modelId', true);
                validator_1.isValidType(data, 'object', 'data', true);
            }, error_1.CatalystZiaError);
            const request = {
                method: REQ_METHOD.post,
                path: `/ml/automl/model/${modelId}`,
                data,
                type: 'json',
                catalyst: true,
                track: true
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
}
exports.Zia = Zia;
